diff --git a/libavcodec/mediacodecdec_common.c b/libavcodec/mediacodecdec_common.c
index 9a627c0..5854a26 100644
--- a/libavcodec/mediacodecdec_common.c
+++ b/libavcodec/mediacodecdec_common.c
@@ -778,6 +778,10 @@ static int mediacodec_dec_get_video_codec(AVCodecContext *avctx, MediaCodecDecCo
     }

     s->codec_name = ff_AMediaCodecList_getCodecNameByType(mime, profile, 0, avctx);
+    if (!s->codec_name && (avctx->hwaccel_flags & AV_HWACCEL_FLAG_ALLOW_PROFILE_MISMATCH)) {
+        profile = -1;
+        s->codec_name = ff_AMediaCodecList_getCodecNameByType(mime, profile, 0, avctx);
+    }
     if (!s->codec_name) {
         // getCodecNameByType() can fail due to missing JVM, while NDK
         // mediacodec can be used without JVM.
